### DISCUSSION TITLE
What does maxConcurrency actually do?

### DISCUSSION THREAD

**Discussion by akhil-dh1** [confluentinc/parallel-consumer#601] [Q&A]
Wanted to know is the maxConcurrency property basically the size of the thread pool that executes the messages parallelly? If so, is it a fixed thread pool or dynamic one? Because in some examples I saw maxConcurrency set to 1000, if that is thread count of a fixed pool, that would crash the system. Just wanted some clarity on this. Thanks.

**Comment by astubbs** [comment]
@akhil-dh1 - in the core module, yes: `maxConcurrency` is the size of the worker pool, and the reason 1000 does not crash anything is what the threads do.

- **What it is.** A fixed `ThreadPoolExecutor` with core and maximum size both set to `maxConcurrency`. Threads are created on demand as work arrives, up to that count, and kept once created; they are ordinary platform threads.
- **Why 1000 is fine for IO.** The threads spend nearly all their time parked on an HTTP call or a database write, so 1000 of them cost stack memory and little else. That is the case the default and the examples are written for, and why the javadoc says to set it much higher than the core count.
- **Why 1000 is wrong for CPU work.** If your function is pure computation, 1000 threads on eight cores is context-switch cost, not throughput; set it near the core count instead.
- **It also sizes the buffer.** The in-flight target is `maxConcurrency x batchSize x load factor`, and the load factor grows under lag, so `maxConcurrency` decides how many records PC holds in memory as well as how many it processes at once. `messageBufferSize` fixes that number if you want it independent (README, "Ordered by Partition" has the tuning notes).
- **In the Vert.x, Reactor and Mutiny modules** it is a ceiling on outstanding asynchronous work, not a thread count.

Where the fork stands: unchanged on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0), and the README only said "the maximum number of concurrent processing operations", which is what left your question open; the README's setup callout now says all of the above (astubbs/parallel-consumer#516, landing). Two open pull requests on the fork, neither merged yet, change the picture: virtual threads for the user function on JDK 21 and later (astubbs/parallel-consumer#360), which makes the thread count stop mattering, and self-tuning concurrency, where the engine finds its own sustainable number instead of asking you to guess (astubbs/parallel-consumer#333).

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
